tools/kafka/commands.go

```diff
@@ -0,0 +1,314 @@
+	if originKey == nil || len(originKey) == 0 {
+		return nil
+	}
+	return sarama.StringEncoder(string(originKey))
+}
+
+func purgeTopic(c *cli.Context) {
+	if !c.IsSet(FlagOffsets) {
+		ErrorAndExit("", fmt.Errorf("target topic name must be provided by flag %v", FlagTopic))
+	}
+
+	offsetStr := c.String(FlagOffsets)
+	offsets, err := parseOffsetStr(offsetStr)
+	if err != nil {
+		ErrorAndExit("input offset is not valid", err)
+	}
+
+	setTopicOffsets(c, offsets, true)
+}
+
+func setTopicOffsets(c *cli.Context, offsetPerPartition map[int32]int64, isForward bool) {
+	if !c.IsSet(FlagTopic) {
+		ErrorAndExit("", fmt.Errorf("target topic name must be provided by flag %v", FlagTopic))
+	}
+	topic := c.String(FlagTopic)
+	if !c.IsSet(FlagConsumerGroup) {
+		ErrorAndExit("", fmt.Errorf("target consumer group name must be provided by flag %v", FlagConsumerGroup))
+	}
+	group := c.String(FlagConsumerGroup)
+
+	cluster := c.String(FlagCluster)
+	clusterFile := c.GlobalString(FlagKafkaClusterFile)
+	useLocal := c.GlobalBool(FlagLocal)
+	brokers := map[string][]string{}
+	var err error
+	if useLocal {
+		brokers = map[string][]string{
+			cluster: {localKafka},
+		}
+	} else {
+		// if not using local mode, cluster names and host file must be provided
+		if len(clusterFile) == 0 || len(cluster) == 0 {
+			ErrorAndExit("", fmt.Errorf("cluster name and host file must be provided must be provided by flags %v,%v", FlagCluster, FlagKafkaClusterFile))
+		}
+		brokers, err = loadClusterConfig(clusterFile, []string{cluster})
+		if err != nil {
+			ErrorAndExit("failed to load cluster from file", err)
+		}
+	}
+
+	config := saramacluster.NewConfig()
+	config.Group.Mode = saramacluster.ConsumerModePartitions
+	consumer, err := saramacluster.NewConsumer(brokers[cluster], group, []string{topic}, config)
+
+outloop:
+	for {
+		select {
+		case <-time.After(heartBeat):
+			_, ok := consumer.Subscriptions()[topic]
+			if !ok {
+				fmt.Println("heartbeat: waiting for subscriptions...")
+				continue
+			}
+			fmt.Println("subs:", consumer.Subscriptions())
+			break outloop
+		}
+	}
+
+	for p, off := range offsetPerPartition {
+		fmt.Printf("partition %v setting to offset %v \n", p, off)
+		if isForward {
+			consumer.MarkPartitionOffset(topic, p, off, "")
+		} else {
+			consumer.ResetPartitionOffset(topic, p, off, "")
+		}
+	}
+
+	if err := consumer.CommitOffsets(); err != nil {
+		ErrorAndExit("failed to commit offsets", err)
+	}
+
+	fmt.Printf("mark offset for topic %v group %v is completed\n", topic, group)
+}
+
+func resetTopic(c *cli.Context) {
+	if !c.IsSet(FlagOffsets) {
+		ErrorAndExit("", fmt.Errorf("target topic name must be provided by flag %v", FlagTopic))
+	}
+	offsetStr := c.String(FlagOffsets)
+	offsets, err := parseOffsetStr(offsetStr)
+	if err != nil {
+		ErrorAndExit("input offset is not valid", err)
+	}
+	setTopicOffsets(c, offsets, false)
+}
+
+func parseOffsetStr(offsetStr string) (map[int32]int64, error) {
```

Comment:
we should be able to get the current water mark and use it for merge and purge instead of manually providing the offset